multi_bot/userver-0.8.0/userver.1
.\"
.\"    userver -- (pronounced you-server or micro-server).
.\"    This file is part of the userver, a high-performance web server designed for
.\"    performance experiments.
.\"
.\"    Copyright (C) 2005-2010 Tim Brecht
.\"    Based on the file originally Copyright (C) 2004  Hewlett-Packard Company
.\"
.\"    Authors: Tim Brecht <brecht@cs.uwaterloo.ca>
.\"    See AUTHORS file for list of contributors to the project.
.\"  
.\"    This program is free software; you can redistribute it and/or
.\"    modify it under the terms of the GNU General Public License as
.\"    published by the Free Software Foundation; either version 2 of the
.\"    License, or (at your option) any later version.
.\"  
.\"    This program is distributed in the hope that it will be useful,
.\"    but WITHOUT ANY WARRANTY; without even the implied warranty of
.\"    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\"    General Public License for more details.
.\"  
.\"    You should have received a copy of the GNU General Public License
.\"    along with this program; if not, write to the Free Software
.\"    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
.\"    02111-1307 USA
.\"
.\"---------------------------------------------------------------------
.TH userver 1 "30 January 2011" "userver-0.8.0"
.IX userver
.\"---------------------------------------------------------------------
.SH NAME
userver \- Micro-server
.br
(A fairly simple web server designed for experimentation)
.\"---------------------------------------------------------------------
.SH SYNOPSIS
.B userver

.RB [ -A | --accepts-only | --close-after-accept ]
.RB [ -B | --block-for-send_events ]
.RB [ -C | --caching-on ]
.RB [ -D | --debug-mask
.IR [0x]N ]
.RB [ -E | --send-events ]
.RB [ -F | --reset-on-close ]
.RB [ -G | --use-getpid ]
.RB [ -H | -h | -? | --help ]
.RB [ -I | --rcv-sock-low-wat
.IR N ]
.RB [ -J | --snd-sock-low-wat
.IR N ]
.RB [ -K | --kdebug-mask
.IR [0x]N ]
.RB [ -L | --full-read
.IR N ]
.RB [ -M | --multi-accept ]
.RB [ -N | --send-loop ]
.RB [ -O | --select-timeout
.IR N ]
.RB [ -P | --count-sigpipes ]
.RB [ -Q | --ecb-hi-water
.IR N ]
.RB [ -R | --read-sockbuf-size
.IR N ]
.RB [ -S | --dont-intr-select ]
.RB [ -T | --port
.IR N ]
.RB [ -U | --auto-accept ]
.RB [ -V | --use-sendfile ]
.RB [ --auto-accept-aperture ]
.RB [ -W | --write-sockbuf-size
.IR N ]
.RB [ -Y | --send-io-events ]
.RB [ -Z | --content-type ]
.RB [ -a | --async-mode ]
.RB [ -b | --ecb-buf-entries
.IR N ]
.RB [ -c | --max-conns
.IR N ]
.RB [ -d | --delay
.IR N ]
.RB [ -e | --extra-accepts
.IR N ]
.RB [ -f | --max-fds
.IR N ]
.RB [ -g | --accept-on-close ]
.RB [ -i | --interactive ]
.RB [ -j | --kernel-info ]
.RB [ -k | --sigio-use-procmask ]
.RB [ -l | --listenq
.IR N ]
.RB [ -m | --accept-count
.IR N ]
.RB [ -n | --no-accepts | --listen-only ]
.RB [ -o | --process-fd-order " [" up | down | writes-down | writes-up | lru | lifo | fifo ] ]
.RB [ -p | --procs
.IR N ]
.RB [ -q | --ecb-low-water
.IR N ]
.RB [ -r | --eager-reads ]
.RB [ -s | --conns-in-server-loop ]
.RB [ -t | --free-fd-threshold
.IR N ]
.RB [ -u | --track-max-fd
.IR N ]
.RB [ -v | --use-poll ]
.RB [ -w | --eager-writes ]
.RB [ -x | --sigio-accepts ]
.RB [ -y | --memcpy ]
.RB [ -z | --rejection-rate ]
.RB [ -1 | --use-epoll ]
.RB [ -2 | --use-epoll2 ]
.RB [ -3 | --use-epoll-ctlv ]
.RB [ --cfg-filename
.IR path ]
.RB [ --close-after-sock-init ]
.RB [ --close-after-read ]
.RB [ --close-after-parse ]
.RB [ --send-polls-for-accepts ]
.RB [ --cache-table-size
.IR N ]
.RB [ --cache-max-bytes
.IR N ]
.RB [ --cache-max-file-size
.IR N ]
.RB [ --cache-max-load-factor
.IR N ]
.RB [ --cache-lock-pages ]
.RB [ --cache-table-print ]
.RB [ --cache-for-spec ]
.RB [ --cache-warm=path ]
.RB [ --trace-summary ]
.RB [ --trace-summary-only ]
.RB [ --use-tcp-cork ]
.RB [ --use-madvise ]
.RB [ --use-accept-send ]
.RB [ --ip-addr
.IR N.N.N.N[:P] ]
.RB [ --use-aio-accept ]
.RB [ --use-aio-read ]
.RB [ --use-aio-write ]
.RB [ --use-aio-close ]
.RB [ --use-socket-aio ]
.RB [ --use-aio-wait ]
.RB [ --aio-read-before-accept ]
.RB [ --aio-read-before-write ]
.RB [ --aio-write-events-limit
.IR N ]
.RB [ --aio-read-events-limit
.IR N ]
.RB [ --aio-accept-events-limit
.IR N ]
.RB [ --aio-accept-thold
.IR N ]
.RB [ --aio-complq-count
.IR N ]
.RB [ --pid-filename
.IR path ]
.RB [ --trace-filename
.IR path ]
.RB [ --read-buffer-size
.IR N ]
.RB [ --reply-buffer-size
.IR N ]
.RB [ --dyn-buffer-size
.IR N ]
.RB [ --num-dyn-buffers
.IR N ]
.RB [ --num-dyn-buffers-per-app
.IR N ]
.RB [ --num-dyn-buffers-per-appserver
.IR N ]
.RB [ --dyn-lock-pages ]
.RB [ --dyn-touch-pages ]
.RB [ --stats-interval
.IR N ]
.RB [ --ignore-fd-setsize ]
.RB [ --use-cpu-mask
.IR [0x]N ]
.RB [ --idle-threshold
.IR N ]
.RB [ --doc-root
.IR path ]
.RB [ --hostname
.IR name ]
.RB [ --app
.IR uri,type[,path[,count]] ]
.RB [ --start-app-server
.IR path[,count][=exec_string] ]
.RB [ --app-req-queue-size
.IR N ]
.RB [ --skip-header ]
.RB [ --victim
.IR filepath ]
.RB [ --victim-skip
.IR N]
.RB [ --cache-miss-skip
.IR N]
.RB [ --call-stats
.IR N]

.\"---------------------------------------------------------------------
.SH DESCRIPTION
.B userver
is a micro web server that is meant to be used to experiment with
the design and implementation of web servers.
In particular the original intention was to permit experimental comparisons
of different event dispatch mechanisms within the same application
framework.
For this reason there are lots and lots and lots of parameters.
They are mainly used to control the behaviour of the server.

Some of the design philosophy and experimental results obtained
using this server are described in [Brecht-2001].
One of note is
that all data structures are allocated at initialization
and are not resized (grown).
This helps to ensure that experiments can be repeated because
with dynamic sizing of data structures
different loads could result in different server behaviour.

The userver uses an event driven architecture.
The idea is to run one event driven server per processor.
If using sendfile on Linux the file system buffer cache will
be shared across all servers. For high-performance one would
want to use sendfile since it doesn't involve extra copying
from the file buffer cache to socket buffers (i.e., it's a zero copy
implementation).


.B IMPORTANT:
This micro web server is not meant to be a full-blown web server.
It is specifically designed for conducting performance experiments.
For that reason design decisions were made in favour of
producing repeatable experiments.

.\"---------------------------------------------------------------------
.SH EXAMPLES

.TP
userver
This starts the
.B userver
with all of its default options.

.TP
userver --max-conns 300 --accept-count 25
This command causes the
.B userver
server to limit
the number of simultaneous connections to 300.
When an event indicating that a new connection is requested
accept is called continuously until
either 25 new connections are accepted,
or there are no more outstanding connections.

.TP
userver --max-conns 300 --accept-count 0
This command causes the
.B userver
server to limit
the number of simultaneous connections to 300.
When an event indicating that a new connection is requested
accept is called repeatedly until
there are no more outstanding connections.
(Note that --accept-count=0 means an infinite number of connections,
i.e., no limits).

.TP
userver --ip-addr 192.168.10.105 --max-conns 300 --listenq 128
.TP
userver --ip-addr 192.168.20.105 --max-conns 300 --listenq 128
This will start up two copies of the userver each accepting
connections from a different interface.

.TP
userver --ip-addr 192.168.10.105:6801 --max-conns 300 --listenq 128
.TP
userver --ip-addr 192.168.20.105:6802 --max-conns 300 --listenq 128
Same as above except one listens on port 6801 and the other on
port 6802.

.TP
userver --port 6800 --max-conns 300 --listenq 128
This command causes
.B userver
to listen for incoming connection requests on port 6800.
It uses a listen queue of length 128 and
it will permit up to 300 simultaneous connections.
Once there are 300 open connections it will not accept new
connections until an existing connection is closed.

.TP
userver --port 6800 --max-conns 300 --listenq 128 --use-poll
This is the same as above except using
.I select()
calls to get events the
.B userver
will use
.I poll().

.TP
userver --port 6800 --max-conns 300 --listenq 128 --use-epoll
This is the same as above except using
.I poll()
calls to get events the
.B userver
will use
.I epoll().

.TP
userver -C --cache-table-size=10000 --cache-max-bytes=262144000 --cache-max-file-size=1048576
Caching is turned on.
Caching parameters limit the number of files that are cached to 10000,
the total number of bytes cached to 250 MB,
and the size of the largest file that can be cached to 1 MB.

.\"---------------------------------------------------------------------
.SH OPTIONS
.rs
.sp 1
The operation of
.B userver
can be controlled through a number of options.
The tool supports both
short (one-character) and long (arbitrary-length) option names.
Short options are prefixed with a single leading dash (-),
long options with a double-dash (--).
Multiple short options can be grouped together (e.g.,
.RB `` -rw ''
is equivalent to
.RB `` "-r -w" '')
and long options can be abbreviated so long as they remain unique.
Parameters to options can be specified either by following the long
option name with an equal sign and the parameter value (e.g.,
.BR --port=6800 )
or by separating the option name and value with whitespace (e.g.,
.BR "--port 6800" ).

.\"-----------------------------------------------------------------------
.SH OPTIONS (GENERAL)

.TP
.BR --idle-threshold= N
In a situation where the number of open connections has reached the limit
(see --max-conns), the userver will close the least recently used connection
if it has not seen activity in the last N seconds.

.TP
.BR "--use-cpu-mask= [0x]N"
On systems that support the
.I sched_setaffinity()
system call, specifies a mask of schedulable CPUs for the userver process.
If prefixed with '0x', the argument is interpreted as hexadecimal; otherwise
it is interpreted as decimal.  The least significant bit corresponds to the
first logical processor number on the system.

.TP
.BR "-f N "
.TP
.BI --max-fds= N
The maximum number of open file descriptors permitted.
I can't recall if this is actually enforced but it is used
to size internal data structures.
Note that there is a relationship between
.B --max-conns
and
.B --max-fds.
If files are not being cached you could (in the extreme case)
have one fd used for the socket connection and one fd used
for the open file.
So one needs to set
.B --max-fds
= more than 2 x
.B --max-conns.
Note: it should be greater because of some descriptors
that are used up by stdin, out, err, and a few open files, etc.

.TP
.BR "-m N "
.TP
.BI --accept-count= N
Call accept repeatedly until either,
.B --max-conns
is reached,
.B accept()
returns EWOULDBLOCK,
or
.I N
new connections have been accepted.
Note that the value of
.I N
can make a substantial difference in the
performance of event dispatch mechanisms
under heavy loads.
See [Brecht-2001].

.TP
.BR "-T N "
.TP
.BI --port= N
Specify the port number that the server should listen for connections on.

.TP
.BR "-l N"
.TP
.BI --listenq= N
Specify the length of the application's listen queue.
This specifies how many connection requests are allowed to be queued
in the application before the kernel rejects incoming connection
requests.
WARNING: on all version of Linux I've ever seen
(up to and including several 2.6.5) the kernel silently
converts any values of N that are greater than 128 to 128.
For details see sys_listen and the value of SOMAXCONNS.

.TP
.BR -L "[1 | 2]"
.TP
.BR --full-read= "[1 | 2]"
Affects how calling read works.
If this is set to 1
the server will loop on reading
the socket until the read fails.
This mainly happens either because there is nothing left to read because
the other end has been closed or until it would have blocked
(EWOULDBLOCK).

If a value of 1 is used the loop on reads will only occur on
calls to read that occur as a result of an event indicating that the first
read should be done.
If a value of 2 is used the loop on reads occur as above
but if
.B --eager-reads
is also set the server will loop on eager reads
(i.e., those that will be done immediately following a the
acceptance and setup of a connection).

.TP
.BR -H
.TP
.BR -h
.TP
.BR -?
.TP
.BR --help
Print out the usage message.

.TP
.BR "-M N "
.TP
.BI --multi-accept= N
Use the
.I multiaccept()
system call to perform accepts.
NOTE: this was a system call that we added to and experimented
with in Linux.
The idea here is that since we are repeatedly calling
.I accept()
to accept a bunch of connections, why not provide a system
call that allows up to a maximum of
.I N
connections to be accepted in one system call.

.TP
.BR -S
.TP
.BR --dont-intr-select
Disable asynchronous event notification when calling
.I select().
I believe that this was initially
meant to work for SIGIO and for SEND.

.TP
.BR "-d N "
.TP
.BI --delay= N
The idea was to introduce a delay that might simulate
work being done by the server (like some computation and/or
a dynamic request).
I couldn't find a good method for delaying for a specified amount
of time (note that it should be small) with a enough accuracy.

.TP
.BR "-e N "
.TP
.BI --extra-accepts= N
While in the middle of processing events that have been obtained
(e.g., from select, or some other mechanism)
periodically poll for new connection requests.
In this case every time
.I N
descriptors are processed we check for new connection requests.
Note that this open was never used effectively.
I found it easier to use and tune the
.B --accept-count
option.

.TP
.BR -g
.TP
.BR --accept-on-close
When a connection is closed check to see if there are any
outstanding connection requests.
Note: that this only occurs if the maximum number of connections
was reached and now a closed connection will reduce the number
of connections below
.B --max-conns.

.TP
.BR "-o [up|down|writes-down|writes-up|lru|lifo|fifo]"
.TP
.BR "--process-fd-order"
.br
[up|down|writes-down|writes-up|lru|lifo|fifo]
.sp 1
Change the order in which file descriptors are handled.
They are numbered from 0 to
.I N
so the options are:
.RS
.IP up:
from 0 to N
.IP down:
from N to 0
.IP writes-down:
to writes from N to 0, then reads from N to 0
.IP writes-up:
to writes from 0 to N, then reads from 0 to N
.IP lru:
least recently descriptor touched first
.IP lifo:
first descriptor (connection) added to the set first
.IP filo:
first descriptor (connection) added to the set last
.RE

.TP
.BR "-p N "
.TP
.BI --procs= N
Start
.I N
copies of the server.
Note that this option hasn't been tested in a long time
and is almost certain to cause problems.

.TP
.BR -s
.TP
.BR --conns-in-server-loop
Add an additional poll for new connections each time through
the server loop.

.TP
.BR "-t N "
.TP
.BI --free-fd-threshold= N
When the maximum number of connections
.B --max-conns
is reached the server stops checking for incoming
connections.
This options controls when the server starts checking again.
It starts checking again after
.I N
connections can be accommodated before
reaching
.B --max-conns.

.TP
.BR -u
.TP
.BR --track-max-fd
The server currently keeps track of the maximum fd ever used
and uses that as a parameter to calls like select.
This option implements code to dynamically keep track of the
maximum file descriptor currently used.
Note that this requires a bit of extra processing but
it's probably not noticeable, especially if it can
save extra overhead on copying fdsets and processing time
in event mechanisms (e.g., select).

.TP
.BR -r
.TP
.BR --eager-reads
try to eagerly read from new connections.
Call
.I read()
to try to read the request as soon as
the new connection is accepted.
In this mode the server optimistically assumes
that data will be available for reading when a
connection is made on a new socket (or very shortly after).
If the assumption is incorrect,
the read call
simply returns with the error EWOULDBLOCK.
Later the event notification mechanism will
indicate when the socket is readable.

.TP
.BR -w
.TP
.BR --eager-writes
try to eagerly perform writes to new
connections when the response is available.
Both
.B --eager-read
and
.B --eager-write
also try to take advantage of
any potential locality effects by working
on the most recently used file descriptor
and socket.

.TP
.BR -y
.TP
.BR --memcpy
Use the
.B memcpy
library call to copy interest sets before calling select.
This is done rather than doing a straight assignment.
This exists because the size of the interest set could be large and I
don't know how clever a compiler would be about assigning
large data structures.
NOTE: it currently copies the side of an fd_set.
This should be changed to only copy the amount of data required
for the maximum fd of interest.

.TP
.BR -z
.TP
.BR --rejection-rate
Use the
.B rejection-rate
is not yet implemented.
The ideas would be to specify the frequency with which
new connection requests should be rejected.

.TP
.BR -V
.TP
.BR --use-sendfile
Use the
.B sendfile()
call on systems that have it available.
This is only used for uncached files and dynamic content.

.TP
.BR --use-madvise
Use
.I madvise()
calls to provide the kernel with hints that files are being read sequentially
and when they should no longer be cached.

.TP
.BR -v
.TP
.BR --use-poll
Use
.I poll()
instead of
.I select()
to get events.

.TP
.BR -1
.TP
.BR --use-epoll
Use
.I epoll()
instead of
.I select()
to get events.

.TP
.BR -2
.TP
.BR --use-epoll2
Use
.I epoll()
instead of
.I select()
to get events,
but attempt some optimizations to reduce the number of
.I epoll_ctl()
calls.
We found the number of epoll_ctl calls to be excessive
so this options calls epoll_ctl when a new connection
is added and sets the interest to be READ and WRITE.
This means that epoll_wait may return events that we
aren't interested in (e.g., that a socket is writable
even though we are busy reading) but we are exploring
the tradeoffs between the two approaches (--use-epoll
and --use-epoll2).

.TP
.BR -3
.TP
.BR --use-epoll-ctlv
Use
.I epoll_ctlv()
to update events.
This requires the addition of a new system call, epoll_ctlv.
This call permits one to collect a bunch of epoll_ctl calls
and change interest in a bunch of fds at once instead of
having to do one system call (epoll_ctl) per change.

.TP
.BR "--cfg-filename path"
Read additional options from a file.  The format of the file is the same
as the command line, i.e. long and/or short options separated by whitespace
(spaces, tabs, linefeeds).  Lines with a '#' character in the first column
are treated as comments and are ignored.  Config files can use
.B --cfg-filename
to nest other config files.

.TP
.BR "--ip-addr N.N.N.N"
.TP
.BR "--ip-addr N.N.N.N:P"
Specify an IP address to listen for new connections on.
Optionally takes a port number P as well.
The special option --ip-addr 0.0.0.0 specifies that the
userver should listen on any address (this is the default).

.TP
.BR "--pid-filename path"
Sets the name of the file where the userver stores its process ID number
while running.  Specify an empty string ("") to suppress pid file generation.
If this option is not specified, a default value of "userver.pid" is used.

.TP
.BR "--read-buffer-size N"
The number of bytes to allocate at initialization time for
.I read()
operations on each connection.

.TP
.BR "--reply-buffer-size N"
The number of bytes to allocate at initialization time for
.I write()
operations on each connection (for static requests only).


.TP
.BR "--stats-interval N"
Print out some simple stats about related to server performance
at a regularly specified interval.
The interval is specified in seconds.
NOTE: this options overrides the --select-timeout option
in order to be able to print out stats according to the selected interval.

.TP
.BR "--ignore-fd-setsize"
If permitted this causes the userver to
ignore limits placed on the number of open connections
and the maximum values of an open file descriptors
that are required when using select (i.e., the FD_SETSIZE).
This option does not work if --track-max-fd is defined
or if the event mechanism being used is select or SEND.

.TP
.BR -Z
.TP
.BR --content-type
Use heuristics to guess what Content-Type header to return, based on the
uri from the request.  For example, if the uri ended in ".gif", the
userver would assume a content type of "image/gif".

.TP
.BR "--doc-root path"
Set the directory of where to find the documents being
server to the specified directory.
When the document root is specified all requests are assumed
to be relative to the specified directory.

.TP
.BR "--hostname name"
Set the name of the machine that the userver is running on.
By default, userver uses
.I gethostname()
to determine the hostname.
The hostname is only used to set the SERVER_NAME CGI parameter when
fulfilling FastCGI requests.


.TP
.BR "--skip-header"
Assume that all files being served contain a header in the
actual file so there is no need for the server to generate
and send a header.
Useful for looking at the impact of things like cork/uncork
when used with sendfile
(i.e., if the header is in the file it's one sendfile system
call instead of cork, write, sendfile, uncork).

.TP
.BR "--call-stats N"
Track statistics for up to N calls/requests.
This is designed to read the Client-Id: header that
we have httperf send in order to correlate what is happening
with requests from different clients on different servers.

.\"-----------------------------------------------------------------------
.SH OPTIONS (FOR CONTROLLING SOCKETS)
.rs
.sp 1
.TP
.BR -F
.TP
.BR --reset-on-close
Calls
.I shutdown()
instead of
.I close().
.br
WARNING do not use this for a real system!!!

.TP
.BR "-I N"
.TP
.BI --rcv-sock-low-wat= N
Sets the sockets receive buffer's low water mark to
.I N.
Not available on all systems.

.TP
.BR -J N
.TP
.BR --snd-sock-low-wat= N
Sets the sockets send buffer's low water mark to
.I N.
Not available on all systems.

.TP
.BR "-O N "
.TP
.BI --select-timeout= N
Set the timeout option to the
.I select()
system call.

.TP
.BR "-R N "
.TP
.BI --read-sockbuf-size= N
.TP
.BR "-W N "
.TP
.BI --write-sockbuf-size= N
Set the size of the read or write socket buffers.
Note that increasing these may help significantly for large
transfers.

.TP

multi_bot/userver-0.8.0/userver.c
/*
    userver -- (pronounced you-server or micro-server).
    This file is part of the userver, a high-performance web server designed for
    performance experiments.

    Copyright (C) 2005-2010 Tim Brecht
    Based on the file originally Copyright (C) 2004  Hewlett-Packard Company

    Authors: Tim Brecht <brecht@cs.uwaterloo.ca>
             David Pariag <db2paria@cs.uwaterloo.ca>
    See AUTHORS file for list of contributors to the project.
  
    This program is free software; you can redistribute it and/or
    modify it under the terms of the GNU General Public License as
    published by the Free Software Foundation; either version 2 of the
    License, or (at your option) any later version.
  
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
    General Public License for more details.
  
    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA
*/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <sys/stat.h>

#include "info.h"
#include "debug.h"
#include "options.h"
#include "lru.h"
#include "q.h"
#include "netstat.h"
#include "rusage.h"
#include "interest_set.h"
#include "sock_listener.h"
#include "sock_special.h"
#include "aio_layer.h"
#include "sigstuff.h"
#include "util.h"

#ifdef SEND
#include "send.h"
#endif /* SEND */

#define PROG_NAME_LEN  (100)

char prog_name[PROG_NAME_LEN + 1];

void print_extra_info();
void do_proc_info(char *proc_file);
void select_loop_init(void);
void poll_loop_init(void);
void send_loop_init(void);
#ifdef HAVE_EPOLL
void epoll_loop_init(void);
#endif /* HAVE_EPOLL */
void aio_loop_init(void);

#ifdef SEND
extern unsigned int send_readfds;
#endif /* SEND */

#ifdef CACHE_MAPPED_NEW
extern void cache_warm(const char *files, const char *docroot, int printcache);
#endif

int
main(int argc, char **argv)
{
  const char *ptr;
  int server_mode = -1;
  int i;
  int rc = 0;
  FILE *xmlf;

#if !defined(HPUX)
  setlinebuf(stdout);
#endif

  ptr = strrchr(argv[0], '/');
  if (ptr == NULL) {
    strncpy(prog_name, argv[0], PROG_NAME_LEN);
  } else {
    strncpy(prog_name, ptr + 1, PROG_NAME_LEN);
  }

  pagesize = getpagesize();
  printf("pagesize = %d\n", pagesize);
  cache_line_size = OPT_CACHE_LINE_EST;

#ifdef TBB_KINFO
  int kinfo_fd = 0;
#endif /* TBB_KINFO */
#ifdef TBB_KINFO
  if (options.kinfo) {
    kinfo_fd = pre_kinfo_init();
  }
#endif /* TBB_KINFO */

  init_print_time(&tnow, &tprev);

  /* must happen before processing arguments */
  options_set();

  /* Process command line arguments */
  process_args(argc, argv);

  /* Check for conflicts in the specified cmd line options */
  check_options();

	printf("INIT START\n");

  num_idle = maxconns;
  printf("main: num_idle = %d\n", num_idle);
  printf("main: maxconns = %d\n", maxconns);
  printf("main: max_fds = %d\n", max_fds);

#ifdef DEBUG_ON
  printf("Debugging is on debugging messages mask = 0x%x\n", MSG_MASK);
#else
  printf("Compiled without debugging\n");
#endif /* DEBUG_ON */

  if (options.use_socket_aio) {
    // for now leave some room for a few open files and
    // a few listening sockets (unfortunately we need to call this before
    // we know how many listening sockets there will be
    // max_sock_sd = maxconns + 20;

    // For now we just pick a value that _SHOULD_ avoid
    // collisions with future tests.
    // Note that if there were sd/fd collisions things
    // _should_ likely work just fine but this may
    // make finding problems easier - since there is
    // potential to control which range of values the
    // underlying aio layer is returning.

#ifdef HAVE_AIO_LAYER
    /* These are just estimates !!! */
    int aio_lowest_sock_sd = OPT_AIO_LOWEST_SD;
    int aio_highest_sock_sd = (maxconns + aio_lowest_sock_sd - 1) + 
                               OPT_AIO_MAX_LISTENERS;
    max_sock_sd = aio_highest_sock_sd;

    if (max_sock_sd > max_fds) {
      printf("main: max_sock_sd = %d > max_fds = %d\n", max_sock_sd, max_fds);
      exit(1);
    }

    printf("main: aio_lowest_sock_sd = %d\n", aio_lowest_sock_sd);
    printf("main: aio_highest_sock_sd = %d\n", aio_highest_sock_sd);
    PRINT_TIME(NOFD, &tnow, &tprev, "main: calling aio_sock_init with "
	"max_sock_sd = %d", max_sock_sd);
    rc = aio_sock_init(aio_lowest_sock_sd, aio_highest_sock_sd);
    if (rc < 0) {
      printf("main: aio_sock_init failed rc = %d\n", rc);
      exit(1);
    }
#endif /* HAVE_AIO_LAYER */

  } else {
    max_sock_sd = max_fds;
  }

  printf("main: max_sock_sd = %d\n", max_sock_sd);
  printf("main: FD_SETSIZE = %d\n", FD_SETSIZE);
  printf("main: sizeof(fd_set) = %d\n", (int) sizeof(fd_set));


#ifdef TBB_KINFO
  if (options.kinfo) {
    kinfo_fd = pre_kinfo_init();
  }
#endif /* TBB_KINFO */

  rc = info_init();


  info_listener_init();

  /* must happen after processing arguments */
  init_app_queues();
  create_pidfile();
  options_valid();
  options_print();

  xmlf = fopen("options.xml", "w");
  if (!xmlf) {
    printf("main: failed to create options.xml\n");
    exit(1);
  }
  options_print_xml(xmlf);
  fclose(xmlf);

#ifndef AIO_WORKAROUND_BUGS
  print_extra_info();
#endif /* AIO_WORKAROUND_BUGS */

#ifdef HAVE_NETSTAT
  netstat_init();
#endif /* HAVE_NETSTAT */

  sock_special_init();

  /* initialize the event dispatch mechanism */
  if (options.use_poll) {
    PRINT_TIME(NOFD, &tnow, &tprev, "userver: calling poll_loop_init");
    poll_loop_init();
#ifdef HAVE_EPOLL
  } else if (options.use_epoll || options.use_epoll2) {
    PRINT_TIME(NOFD, &tnow, &tprev, "userver: calling epoll_loop_init");
    epoll_loop_init();
#endif /* HAVE_EPOLL */
#ifdef HAVE_AIO_LAYER
  } else if (options.use_socket_aio) {
    PRINT_TIME(NOFD, &tnow, &tprev, "userver: calling aio_loop_init");
    aio_loop_init();
#endif /* HAVE_AIO_LAYER */
  } else if (!options.send_loop) {
    assert(options.send_loop == 0);
    PRINT_TIME(NOFD, &tnow, &tprev, "userver: calling select_loop_init");
    select_loop_init();
  } else {
#ifdef SEND
    assert(options.send_loop == 1);
    PRINT_TIME(NOFD, &tnow, &tprev, "userver: calling send_loop_init");
    send_loop_init();
#endif
  }

  switch (options.get_connections)
  {
    case OPT_CONN_WITH_SELECT_POLL_EPOLL:
      server_mode = LISTENER_NOT_ASYNC_INIT;
      break;

#ifdef SEND
    case OPT_CONN_WITH_SEND_SELECT:
      server_mode = LISTENER_NOT_ASYNC_INIT | LISTENER_SEND_INIT;
      break;
#endif /* SEND */
    
    case OPT_CONN_WITH_SIGIO:
      server_mode = LISTENER_DO_ASYNC_INIT;
      break;

    case OPT_CONN_WITH_SEND_EVTS:
      server_mode = LISTENER_SEND_INIT;
      break;

    case OPT_CONN_WITH_AIO_ACCEPT:
      server_mode = LISTENER_AIO_INIT;
      break;

    default :
      printf("%s: options.get_connections = %d not handled\n",
              argv[0], options.get_connections);
      exit(1);
      break;
  } /* switch */

  printf("Calling server_init with mode = 0x%x = %d\n", 
          server_mode, server_mode);
  server_init(server_mode);

  for (i=sock_listener_min; i<=sock_listener_max; i++) {
    if (sock_is_listener(i)) {
      printf("listen_sd = %d\n", i);
      if (i > max_sd) {
        max_sd = i;
      }
    }
  }

  switch (options.process_sds_order)
  {
      case OPT_PROCESS_SDS_LRU:
	lru_copy_init();
	break;
    
      case OPT_PROCESS_SDS_LIFO:
        q_init(max_fds+1);
	if (options.get_connections == OPT_CONN_WITH_SELECT_POLL_EPOLL) {
	  for (i=sock_listener_min; i<=sock_listener_max; i++) {
	    if (sock_is_listener(i)) {
	      q_add_to_front(i);
	    }
	  }
	}
	break;

      case OPT_PROCESS_SDS_FIFO:
        q_init(max_fds+1);
	if (options.get_connections == OPT_CONN_WITH_SELECT_POLL_EPOLL) {
	  for (i=sock_listener_min; i<=sock_listener_max; i++) {
	    if (sock_is_listener(i)) {
	      q_add_to_rear(i);
	    }
	  }
	}
	break;

      default :
	/* do nothing */
	break;
  }

  if (options.caching_on) {
    initCache(options.cache_table_size, options.cache_max_bytes, 
              options.cache_max_file_size, options.cache_max_load_factor, 
              options.cache_lock_pages);
#ifdef CACHE_MAPPED_NEW
      if (options.cache_warm) {
        cache_warm(options.cache_warm_file,
                   (options.doc_root[0] != '\0') ? options.doc_root : NULL,
                   options.cache_table_print);
       }
#endif
  }

  rusage_init();

  fork_servers(numprocs);
  return 0;
}

/* ---------------------------------------------------------------------- */
void 
print_extra_info()
{
  int rc = 0;
#ifdef HAVE_IFCONFIG
  printf("ifconfig output:\n");
  fflush(stdout);
  fflush(stderr);
  rc = system("/sbin/ifconfig");
  fflush(stdout);
  fflush(stderr);
  printf("-----------------------------------------------\n");
#endif /* HAVE_IFCONFIG */

#ifdef HAVE_PROC_FILE_INFO
  do_proc_info("/proc/version");
  do_proc_info("/proc/sys/kernel/rtsig-max");
  do_proc_info("/proc/sys/fs/file-max");
  do_proc_info("/proc/sys/net/ipv4/tcp_max_syn_backlog");
  do_proc_info("/proc/sys/net/ipv4/ip_local_port_range");
  do_proc_info("/proc/sys/net/ipv4/tcp_tw_recycle");
  do_proc_info("/proc/sys/net/ipv4/tcp_max_tw_buckets");
  do_proc_info("/proc/sys/net/ipv4/tcp_fin_timeout");
  do_proc_info("/proc/sys/net/ipv4/tcp_abort_on_overflow");
  do_proc_info("/proc/sys/net/core/somaxconn");

  fflush(stdout);
  fflush(stderr);
  printf("From /proc/cpuinfo:\n");
  fflush(stdout);
  fflush(stderr);
  rc = system("egrep 'model name|processor|MHz|cache|_id' /proc/cpuinfo");
  fflush(stdout);
  fflush(stderr);

  printf("From /proc/interrupts (before):\n");
  rc = system("grep 'eth' /proc/interrupts");

#endif /* HAVE_PROC_FILE_INFO */

#ifdef HAVE_LIMITS
  printf("-----------------------------------------------\n");
  printf("Limits:\n");
  fflush(stdout);
  fflush(stderr);
  rc = system("csh -c limit");
  fflush(stdout);
  fflush(stderr);
  printf("-----------------------------------------------\n");
  fflush(stdout);
  fflush(stderr);
#endif /* HAVE_LIMITS */
}
/* ---------------------------------------------------------------------- */
void
do_proc_info(char *proc_file)
{
  char cmd[200];
  struct stat st;
  int rc;

  rc = stat(proc_file, &st);
  if (rc == 0) {
    printf("%s   ", proc_file);
    /* Interestingly these are required in order to get proper output */
    fflush(stdout);
    fflush(stderr);
    sprintf(cmd, "cat %s", proc_file);
    rc = system(cmd);
    fflush(stdout);
    fflush(stderr);
  }
}

/* ---------------------------------------------------------------------- */

void
create_pidfile()
{
  FILE *fp;

  printf("%s pid = %d\n", prog_name, (int) getpid());

  if (options.pid_filename[0] != '\0') {
    fp = fopen(options.pid_filename, "w");
    if (fp == NULL) {
      perror("create_pidfile: fopen failed");
      exit(1);
    }
    fprintf(fp, "%d\n", (int) getpid());
    fclose(fp);
  }
}

/* ---------------------------------------------------------------------- */

void
delete_pidfile()
{
  if (options.pid_filename[0] != '\0') {
    unlink(options.pid_filename);
  }
}
